experimental/cuda2/pipeline_layout.c

```diff
@@ -0,0 +1,238 @@
+
+//===----------------------------------------------------------------------===//
+// iree_hal_cuda2_descriptor_set_layout_t
+//===----------------------------------------------------------------------===//
+
+typedef struct iree_hal_cuda2_descriptor_set_layout_t {
+  // Abstract resource used for injecting reference counting and vtable;
+  // must be at offset 0.
+  iree_hal_resource_t resource;
+
+  // The host allocator used for creating this descriptor set layout struct.
+  iree_allocator_t host_allocator;
+
+  // The total number of bindings in this descriptor set.
+  iree_host_size_t binding_count;
+} iree_hal_cuda2_descriptor_set_layout_t;
+
+static const iree_hal_descriptor_set_layout_vtable_t
+    iree_hal_cuda2_descriptor_set_layout_vtable;
+
+static iree_hal_cuda2_descriptor_set_layout_t*
+iree_hal_cuda2_descriptor_set_layout_cast(
+    iree_hal_descriptor_set_layout_t* base_value) {
+  IREE_HAL_ASSERT_TYPE(base_value,
+                       &iree_hal_cuda2_descriptor_set_layout_vtable);
+  return (iree_hal_cuda2_descriptor_set_layout_t*)base_value;
+}
+
+iree_status_t iree_hal_cuda2_descriptor_set_layout_create(
+    iree_hal_descriptor_set_layout_flags_t flags,
+    iree_host_size_t binding_count,
+    const iree_hal_descriptor_set_layout_binding_t* bindings,
+    iree_allocator_t host_allocator,
+    iree_hal_descriptor_set_layout_t** out_descriptor_set_layout) {
+  IREE_ASSERT_ARGUMENT(!binding_count || bindings);
+  IREE_ASSERT_ARGUMENT(out_descriptor_set_layout);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  *out_descriptor_set_layout = NULL;
+
+  iree_hal_cuda2_descriptor_set_layout_t* descriptor_set_layout = NULL;
+  iree_status_t status =
+      iree_allocator_malloc(host_allocator, sizeof(*descriptor_set_layout),
+                            (void**)&descriptor_set_layout);
+
+  if (iree_status_is_ok(status)) {
+    iree_hal_resource_initialize(&iree_hal_cuda2_descriptor_set_layout_vtable,
+                                 &descriptor_set_layout->resource);
+    descriptor_set_layout->host_allocator = host_allocator;
+    descriptor_set_layout->binding_count = binding_count;
+    *out_descriptor_set_layout =
+        (iree_hal_descriptor_set_layout_t*)descriptor_set_layout;
+  }
+
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
+
+iree_host_size_t iree_hal_cuda2_descriptor_set_layout_binding_count(
+    iree_hal_descriptor_set_layout_t* base_descriptor_set_layout) {
+  iree_hal_cuda2_descriptor_set_layout_t* descriptor_set_layout =
+      iree_hal_cuda2_descriptor_set_layout_cast(base_descriptor_set_layout);
+  return descriptor_set_layout->binding_count;
+}
+
+static void iree_hal_cuda2_descriptor_set_layout_destroy(
+    iree_hal_descriptor_set_layout_t* base_descriptor_set_layout) {
+  iree_hal_cuda2_descriptor_set_layout_t* descriptor_set_layout =
+      iree_hal_cuda2_descriptor_set_layout_cast(base_descriptor_set_layout);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_allocator_free(descriptor_set_layout->host_allocator,
```

Comment:
prefer capturing the host_allocator locally
```suggestion
  iree_allocator_t host_allocator = descriptor_set_layout->host_allocator;
  iree_allocator_free(host_allocator,
```
this makes it clearer that the lifetime of the allocator value is longer than the thing you're freeing - it's technically fine but always looks sketchy to pass a field value on something you're freeing :)

Reply:
Good point!